# Purpose of notebook: to make txt files with sources and text body into csv files and merge them 

In [15]:
# import/load packages

## install
%pip install pandas
%pip install chardet
%pip install cchardet


## import
import os
import pandas as pd
import chardet
import cchardet
import re
from bs4 import BeautifulSoup

Note: you may need to restart the kernel to use updated packages.
Note: you may need to restart the kernel to use updated packages.
Note: you may need to restart the kernel to use updated packages.


## From TXT to CSV (+ global preprocessing) - sources.txt + text.txt


### sources.txt

In [16]:
# define file path
file_path = "/work/Bachelor/sample_data/original_sample/sources.txt"

# define column names
column_names = ["textID", "words", "date", "country", "source", "url", "title"]

# read the file into DataFrame and skip the first two lines
df_sources_original_sample = pd.read_csv(
    file_path,
    delimiter='\t',
    encoding='ISO-8859-1',
    skiprows=2,  # skip the first two lines (header and separator line)
    names=column_names  # assign new column names while reading the file
)

df_sources_original_sample

textID  words      date country                         source  \
0        11241    397  13-01-06      US                         Kotaku   
1        11242    757  13-01-06      US                 Michigan Radio   
2        11243    755  13-01-06      US            New York Daily News   
3        11244   1677  13-01-06      US                 OregonLive.com   
4        21242    794  13-01-11      US                   Ars Technica   
...        ...    ...       ...     ...                            ...   
2955  15641240    332  16-11-29      US  Business Wire (press release)   
2956  15641241    226  16-11-29      US                       Fox News   
2957  15641242    230  16-11-29      US             Firstcoastnews.com   
2958  15641243    193  16-11-29      US           Entertainment Weekly   
2959  15651244   1255  16-11-29      MY            malaysiandigest.com   

                                                    url  \
0     http://kotaku.com/5973495/author-of-the-warrio...   
1     http://michiganradio.org/post/thats-what-they-...   
2     http://www.nydailynews.com/life-style/eats/bes...   
3     http://www.oregonlive.com/performance/index.ss...   
4     http://arstechnica.com/gadgets/2013/01/ask-ars...   
...                                                 ...   
2955  http://www.businesswire.com/news/home/20161129...   
2956  http://www.foxnews.com/politics/2016/11/29/dep...   
2957  http://www.firstcoastnews.com/news/fire-surrou...   
2958  http://www.ew.com/article/2016/11/29/alexis-bl...   
2959  http://www.malaysiandigest.com/sports/645650-j...   

                                                  title  
0     Author of The Warriors, Cult Film Adapted to H...  
1     That's What They Say: Dialect Society chooses ...  
2                           Best of New York: Croissant  
3     Reflecting on a quarter-century of growth in P...  
4     Ask Ars: Does Facebook auto-delete content aft...  
...                                                 ...  
2955  Jungle Disk Appoints Matt Bradley to Board of ...  
2956  Department of Homeland Security overwhelmed by...  
2957  Flames surround Ripley's Aquarium, thousands o...  
2958  Alexis Bledel is working on a Sisterhood of th...  
2959  Jose Mourinho Has Found An Effective Role For ...  

[2960 rows x 7 columns]

### text.txt

In [17]:
# define file path
file_path = "/work/Bachelor/sample_data/original_sample/text.txt"

# open the file and read its contents into a string
with open(file_path, 'r', encoding='utf-8') as file:
    raw_text = file.read()  # Read all text data from the file

sample = raw_text
sample = re.sub(" ([.,?!':])", r"\1", sample) # punctuation characters are all needlessly preceded by a space, remove that space
sample = re.sub("@ @ @ @ @ @ @ @ @ @", "CENSORED", sample) # replace the keyword chosen to mark words or entities that have been censored

# Step 1: Split the text into articles based on '@@' markers
article_ids = re.findall(r"@@(\d+)", sample) # extract list of all article IDs by matching pattern of any sequence of digits following exactly two @
articles = re.split(r'"?@@\d+ ', sample)[1:] # split articles on article IDs, i.e. a sequence of digits preceded by exactly two @s and sometimes a quotation mark preceding that
articles = [art[art.find("<p> ") + 4:].strip().replace(" <p> ", "\n") for art in articles] # for each article, skip to the first paragraph tag to exclude heading/title text, then replace subsequent markers for new paragraphs with line breaks. OBS maybe 'words' column contains incorrect n words because of this. 

print(len(article_ids), len(articles)) # check that we found equally many IDs and article bodies

df_articles_original_sample = pd.DataFrame(data = dict(textID = article_ids, body = articles)) # create dataframe with extracted IDs and corresponding article contents
df_articles_original_sample["textID"] = df_articles_original_sample["textID"].astype(int) # cast textIDs to int to enable merging with `sources` dataframe
df_articles_original_sample.sample(10) # sample ten random rows from the dataframe to examine whether things look as expected


2914 2914


textID                                               body
2360  10351242  THE police authorities have arrested four " dr...
1553   4601243  This is the first time we've seen this prototy...
1138   3321243  Agro-tourism, agriculture education in schools...
1948   8301241  Rawalpindi -- Punjab Governor Malik Muhammad R...
1133   3311242  Novak Djokovic lost a point in bizarre fashion...
1862   7851240  Words by Mad Dog BradleyMarch 23, 2016\nProduc...
604    1661244  Lumosity.com, the leading online provider of g...
2408  13671243  Mamadou Sakho and Divock Origi of Liverpool du...
511    1411241  If successful, England are likely to have to p...
2785  14891240  The Movement for the Actualization of the Sove...

In [18]:
joined_original_sample = df_sources_original_sample.merge(df_articles_original_sample, on = "textID", how = "right") # join sources dataframe with their corresponding article content using the textIDs, keeping only the articles where source information was available
joined_original_sample

textID   words      date country               source  \
0        11241   397.0  13-01-06      US               Kotaku   
1        11242   757.0  13-01-06      US       Michigan Radio   
2        11243   755.0  13-01-06      US  New York Daily News   
3        11244  1677.0  13-01-06      US       OregonLive.com   
4        21242   794.0  13-01-11      US         Ars Technica   
...        ...     ...       ...     ...                  ...   
2909  15281244   405.0  16-11-08      SG           Catch News   
2910  15291240   393.0  16-11-08      CA     Edmonton Journal   
2911  15291242  1000.0  16-11-08      CA        Timmins Press   
2912  15291243   657.0  16-11-08      CA         Windsor Star   
2913  15291244   888.0  16-11-08      CA  CFJC Today Kamloops   

                                                    url  \
0     http://kotaku.com/5973495/author-of-the-warrio...   
1     http://michiganradio.org/post/thats-what-they-...   
2     http://www.nydailynews.com/life-style/eats/bes...   
3     http://www.oregonlive.com/performance/index.ss...   
4     http://arstechnica.com/gadgets/2013/01/ask-ars...   
...                                                 ...   
2909  http://www.catchnews.com/india-news/did-modi-s...   
2910  http://edmontonjournal.com/news/local-news/cou...   
2911  http://www.timminspress.com/2016/11/08/5m-fund...   
2912  http://windsorstar.com/news/local-news/pelissi...   
2913  http://www.cfjctoday.com/article/547992/burrow...   

                                                  title  \
0     Author of The Warriors, Cult Film Adapted to H...   
1     That's What They Say: Dialect Society chooses ...   
2                           Best of New York: Croissant   
3     Reflecting on a quarter-century of growth in P...   
4     Ask Ars: Does Facebook auto-delete content aft...   
...                                                 ...   
2909  Did Modi silently visit Kashmir? Govts PR arm ...   
2910  Council votes for $30 per year Edmonton back a...   
2911      $5M funding boost for six projects in Timmins   
2912  Pelissier Street garage main floor to become p...   
2913  Burrows helps Canucks beat Rangers 5-3 to end ...   

                                                   body  
0     Sol Yurick, the writer whose 1965 novel " The ...  
1     For this week's edition of " That's What They ...  
2     French Tart chef Laurent Chavenet claims to wo...  
3     The highlight of the Portland Center for Perfo...  
4     In 1998, Ask Ars was an early feature of the n...  
...                                                 ...  
2909  Did you know that Prime Minister Narendra Modi...  
2910  Council voted Tuesday to support a new $30 per...  
2911  Minister of the Status of Women Patty Hajdu ( ...  
2912  A closed-door council meeting that reversed th...  
2913  NEW YORK -- The Vancouver Canucks broke throug...  

[2914 rows x 8 columns]

### It works!!
### OBS there are fewer rows in text (2,914) than in sources (2,960), which can be caused by a) text isn't split correctly, b) sources are simply missing from the dataset hence fewer rows. Consider investigating this.